### PR TITLE
refactor: Migrates search index resource and data sources to new SDK

### PR DESCRIPTION
**Pull request by lantoli**
## Description

Jira ticket: [INTMDB-1195](https://jira.mongodb.org/browse/INTMDB-1195)

Migrates search index resource and data sources to new SDK.

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contribution guidelines](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/CONTRIBUTING.md)
- [X] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code
- [x] If changes include deprecations or removals, I defined an isolated PR with a relevant title as it will be used in the auto-generated changelog.

### CODE REVIEW

File: mongodbatlas/data_source_mongodbatlas_search_indexes.go [hunk 1]
```diff
@@ -65,17 +62,14 @@ func dataSourceMongoDBAtlasSearchIndexesRead(ctx context.Context, d *schema.Reso
 		return diag.Errorf("project_id, cluster_name, database and collection_name must be configured")
 	}
 
-	options := &matlas.ListOptions{
-		PageNum:      d.Get("page_num").(int),
-		ItemsPerPage: d.Get("items_per_page").(int),
-	}
```
Comment (lantoli) [inline mongodbatlas/data_source_mongodbatlas_search_indexes.go:71]: still pending what to do with page_num and items_per_page, but please review the rest of the PR

File: mongodbatlas/resource_mongodbatlas_search_index.go [hunk 2]
```diff
@@ -205,26 +202,24 @@ func resourceMongoDBAtlasSearchIndexUpdate(ctx context.Context, d *schema.Resour
 	}
 
 	if d.HasChange("search_analyzer") {
-		searchIndex.SearchAnalyzer = d.Get("search_analyzer").(string)
+		searchIndex.SearchAnalyzer = stringPtr(d.Get("search_analyzer").(string))
 	}
 
+	dynamic := d.Get("mappings_dynamic").(bool)
```
Comment (marcosuma) [inline mongodbatlas/resource_mongodbatlas_search_index.go]: any reason why this outside the if?

File: mongodbatlas/resource_mongodbatlas_search_index.go [hunk 3]
```diff
@@ -334,73 +325,53 @@ func resourceMongoDBAtlasSearchIndexRead(ctx context.Context, d *schema.Resource
...
-}
-
-func marshallSearchIndexMappingsField(fields map[string]interface{}) (string, error) {
-	if len(fields) == 0 {
-		return "", nil
-	}
-
-	mappingFieldJSON, err := json.Marshal(fields)
-	return string(mappingFieldJSON), err
+func marshalSearchIndex(fields any) (string, error) {
```
Comment (marcosuma) [inline mongodbatlas/resource_mongodbatlas_search_index.go:340]: why are we changing from `[]map[string]interface{}` to `any`?

Comment (marcosuma) [reply mongodbatlas/resource_mongodbatlas_search_index.go:340]: maybe better `[]map[string]any`?

File: website/docs/r/search_index.html.markdown [hunk 4]
```diff
@@ -258,10 +258,10 @@ An [Atlas Search analyzer](https://docs.atlas.mongodb.com/reference/atlas-search
 
 * `token_filters` - Array containing zero or more token filters. Always require a type field, and some take additional options as well:
   ```terraform
-  "tokenFilters":{
+  "tokenFilters":[{
```
Comment (marcosuma) [inline website/docs/r/search_index.html.markdown:261]: I guess this PR is gonna be a breaking change?

File: mongodbatlas/resource_mongodbatlas_search_index.go (hunk 3 above)
Comment (lantoli) [reply mongodbatlas/resource_mongodbatlas_search_index.go:340]: this method is just realling calling json.Marshal that accepts any. 
so i joined marshallSearchIndexAnalyzers and marshallSearchIndexMappingsField that accepts an array or a map into the same fuction, no need to have separate functions for this

File: website/docs/r/search_index.html.markdown (hunk 4 above)
Comment (lantoli) [reply website/docs/r/search_index.html.markdown:261]: we agreed with @Zuhairahmed  that this is not a breaking change but a documentation bug.  In fact the title is fine "Array containing zero or more token filters" but the example is wrong

File: mongodbatlas/resource_mongodbatlas_search_index.go (hunk 2 above)
Comment (lantoli) [reply mongodbatlas/resource_mongodbatlas_search_index.go]: no, moving inside, thx.  it must be a variable though so we can take the reference with &

File: mongodbatlas/resource_mongodbatlas_search_index.go (hunk 3 above)
Comment (AgustinBettati) [reply mongodbatlas/resource_mongodbatlas_search_index.go:340]: we could rename the function directly to something more representative, like `toJsonString`

**Review by AgustinBettati — APPROVED** [review]
LGTM

File: mongodbatlas/resource_mongodbatlas_search_index.go (hunk 3 above)
Comment (lantoli) [reply mongodbatlas/resource_mongodbatlas_search_index.go:340]: I like the idea of calling Marshal as there is an opossed method to unmarshal,   and SearchIndex as it's meant to be used in this resource.  so marshal as packaging the info (that turns out to be in json) and unmarshaling to unpackage it

File: mongodbatlas/data_source_mongodbatlas_search_indexes.go (hunk 1 above)
Comment (lantoli) [reply mongodbatlas/data_source_mongodbatlas_search_indexes.go:71]: will be done in another PR to better track changelog

**Pull request merged**